arirs/src/channel.rs

```diff
@@ -34,822 +61,431 @@ pub enum OriginateParams<'a> {
+    pub endpoint: &'a str,
+    pub app: &'a str,
+    #[serde(skip_serializing_if = "<[_]>::is_empty")]
+    #[serde(serialize_with = "join_serialize")]
+    pub app_args: &'a [&'a str],
+    pub channel_id: Option<&'a str>,
+    pub other_channel_id: Option<&'a str>,
+    pub originator: Option<&'a str>,
+    #[serde(skip_serializing_if = "<[_]>::is_empty")]
+    #[serde(serialize_with = "join_serialize")]
+    pub formats: &'a [&'a str],
+}
 
-        reqwest::Client::new().post(url.finish().to_owned()).send().await?;
-
-        event!(Level::INFO, "sent dtmf '{}' to channel with id {}", dtmf, self.id);
-
-        Ok(())
-    }
-
-    #[instrument(level = "debug")]
-    pub async fn mute(&self, client: &Client, direction: Direction) -> Result<()> {
-        let url = client
-            .url
-            .join(&format!("channels/{}/mute", self.id))?
-            .query_pairs_mut()
-            .append_pair("api_key", &client.get_api_key())
-            .append_pair("direction", &format!("{}", direction))
-            .finish()
-            .to_owned();
-
-        reqwest::Client::new().post(url).send().await?;
-        event!(Level::INFO, "muted channel with id {}", self.id);
-        Ok(())
-    }
-
-    #[instrument(level = "debug")]
-    pub async fn unmute(&self, client: &Client, direction: Direction) -> Result<()> {
-        let url = client
-            .url
-            .join(&format!("channels/{}/mute", self.id))?
-            .query_pairs_mut()
-            .append_pair("api_key", &client.get_api_key())
-            .append_pair("direction", &format!("{}", direction))
-            .finish()
-            .to_owned();
-
-        reqwest::Client::new().delete(url).send().await?;
-        event!(Level::INFO, "unmuted channel with id {}", self.id);
-        Ok(())
-    }
-
-    #[instrument(level = "debug")]
-    pub async fn hold(&self, client: &Client) -> Result<()> {
-        let url = client
-            .url
-            .join(&format!("channels/{}/hold", self.id))?
-            .query_pairs_mut()
-            .append_pair("api_key", &client.get_api_key())
-            .finish()
-            .to_owned();
-
-        reqwest::Client::new().post(url).send().await?;
-        event!(Level::INFO, "started hold on channel with id {}", self.id);
-        Ok(())
-    }
-
-    #[instrument(level = "debug")]
-    pub async fn unhold(&self, client: &Client) -> Result<()> {
-        let url = client
-            .url
-            .join(&format!("channels/{}/hold", self.id))?
-            .query_pairs_mut()
-            .append_pair("api_key", &client.get_api_key())
-            .finish()
-            .to_owned();
-
-        reqwest::Client::new().delete(url).send().await?;
-        event!(Level::INFO, "stopped hold on channel with id {}", self.id);
-        Ok(())
-    }
-
-    #[instrument(level = "debug")]
-    pub fn start_moh(&self, _client: &Client) -> Result<()> {
-        unimplemented!()
+impl RequestClient {
```

Comment:
suggestion: if we're doing a refactor of how the different modules work, can't we abstract out the different model functionalities (`bridge`, `channel`, `playback`) to be implemented traits so we don't populate `RequestClient`s internal namespace?

Reply:
Possibly, I was thinking about simply prefixing the methods with `playback_`, `bridge_` etc.

Reply:
I still think that bloats the namespace a bit too much. It does improve the scope of functions but I generally liked the previous interface of `Channel::func(&client)` so I think a `use ChannelExt; client.originate(params);` would be cleaner. I also think this would improve the ability to e.g. mock which could be nice for tests. Though mocks are generally bad :)

Reply:
I've also thought about implementing traits on the client. Problem is that a lot of functions already collide without a prefix (bridge/channel, live/stored recording, listing and getting by id). Prefixes will in many cases be needed anyways, or there will be quite a few cases where disambiguation is required, ex. `<RequestClient as ChannelClient>::stop_moh(&client, params)`, which I think should be avoided as much as possible.

There's probably also plenty of work that can be done do reduce the number of `ResponseClient` methods in the first place. Say by ombining the `_with_id()` methods to their counterparts by setting the id optional. Or say by using generics on the common create, list, get, remove/destroy metods. I think that this should be the primary focus point. The type system should be able to infer a whole lot if we typeset the ids and implement some trait with associated consts and types on them.